File: PrimaMateria/PrimaMateria/Classes/XTRPreferencesViewController.m
//
//  XTRPreferencesViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright 2016 xTrensa. All rights reserved.
//

@interface XTRPreferencesViewController ()
- (void)loadDocument: (NSString *) documentName inView: (UIWebView *) webView;
- (void)loadUserDefaults;
- (void)populateSeriesColors;
- (void)populateShowTransitionsState: (BOOL) aFlag;
- (void)populateElementBubbleState: (BOOL) aFlag;
- (void)populateSplashScreenState: (BOOL) aFlag;
@end

@implementation XTRPreferencesViewController

#pragma mark Private Methods

- (void)colorSelected: (NSNotification *)notification {
    NSDictionary *object = (NSDictionary *)notification.object;
    
    if(object != nil) {
        NSNumber *redComponent = (NSNumber*)object[RED_COLOR_COMPONENT];
        NSNumber *greenComponent = (NSNumber*)object[GREEN_COLOR_COMPONENT];
        NSNumber *blueComponent = (NSNumber*)object[BLUE_COLOR_COMPONENT];
        NSString *seriesColorKey = object[SERIES_COLOR_KEY];
        
        UIColor *aColor = [UIColor colorWithRed:redComponent.floatValue green:greenComponent.floatValue blue:blueComponent.floatValue alpha:1.0];
        NSData *colorData = [NSKeyedArchiver archivedDataWithRootObject:aColor];
        
        [XTRPropertiesStore storeColorData: colorData forColorKey: seriesColorKey];
        
        if ([seriesColorKey isEqualToString: SERIES_ACTINIDE]) {
            self.seriesActinideButton.backgroundColor = aColor;
            [self.seriesActinideButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_ALKALI_EARTH_METAL]) {
            self.seriesAlkaliEarthMetalButton.backgroundColor = aColor;
            [self.seriesAlkaliEarthMetalButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_ALKALI_METAL]) {
            self.seriesAlkaliMetalButton.backgroundColor = aColor;
            [self.seriesAlkaliMetalButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_HALOGEN]) {
            self.seriesHalogenButton.backgroundColor = aColor;
            [self.seriesHalogenButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_LANTHANIDE]) {
            self.seriesLanthanideButton.backgroundColor = aColor;
            [self.seriesLanthanideButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_METAL]) {
            self.seriesMetalButton.backgroundColor = aColor;
            [self.seriesMetalButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_NOBLE_GAS]) {
            self.seriesNobleGasButton.backgroundColor = aColor;
            [self.seriesNobleGasButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_NON_METAL]) {
            self.seriesNonMetalButton.backgroundColor = aColor;
            [self.seriesNonMetalButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_TRANSACTINIDES]) {
            self.seriesTransactinidesButton.backgroundColor = aColor;
            [self.seriesTransactinidesButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        } else if ([seriesColorKey isEqualToString: SERIES_TRANSITION_METAL]) {
            self.seriesTransitionMetalButton.backgroundColor = aColor;
            [self.seriesTransitionMetalButton setTitleColor: [UIColor reverseColorOf: aColor] forState: UIControlStateNormal];
        }
        [[NSNotificationCenter defaultCenter] postNotificationName: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
    }
    
    [self dismissViewControllerAnimated:YES completion: nil];
}

- (void)loadDocument: (NSString *) documentName inView: (UIWebView *) webView {
    NSString *path = [[NSBundle mainBundle] pathForResource: documentName ofType: nil];
    NSURL *url = [NSURL fileURLWithPath: path];
    NSURLRequest *request = [NSURLRequest requestWithURL: url];
    [webView loadRequest: request];
}

- (void)loadUserDefaults {
    self.self.elementBubbleSwitch.on = [XTRPropertiesStore retreiveElementBubblesState];
    self.showTransitionsBubbleSwitch.on = [XTRPropertiesStore retreiveShowTransitionsState];
    self.splashScreenSwitch.on = [XTRPropertiesStore retreiveSplashScreenState];
}

- (void)populateSeriesColors {
    self.seriesActinideButton.backgroundColor = [XTRColorFactory actinideColor];
    self.seriesAlkaliEarthMetalButton.backgroundColor = [XTRColorFactory alkaliEarthMetalColor];
    self.seriesAlkaliMetalButton.backgroundColor = [XTRColorFactory alkaliMetalColor];
    self.seriesHalogenButton.backgroundColor = [XTRColorFactory halogenColor];
    self.seriesLanthanideButton.backgroundColor = [XTRColorFactory lanthanideColor];
    self.seriesMetalButton.backgroundColor = [XTRColorFactory metalColor];
    self.seriesNobleGasButton.backgroundColor = [XTRColorFactory nobleGasColor];
    self.seriesNonMetalButton.backgroundColor = [XTRColorFactory nonMetalColor];
    self.seriesTransactinidesButton.backgroundColor = [XTRColorFactory transactinideColor];
    self.seriesTransitionMetalButton.backgroundColor = [XTRColorFactory transitionMetalColor];
    
    [self.seriesActinideButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory actinideColor]] forState: UIControlStateNormal];
    [self.seriesAlkaliEarthMetalButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory alkaliEarthMetalColor]] forState: UIControlStateNormal];
    [self.seriesAlkaliMetalButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory alkaliMetalColor]] forState: UIControlStateNormal];
    [self.seriesHalogenButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory halogenColor]] forState: UIControlStateNormal];
    [self.seriesLanthanideButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory lanthanideColor]] forState: UIControlStateNormal];
    [self.seriesMetalButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory metalColor]] forState: UIControlStateNormal];
    [self.seriesNobleGasButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory nobleGasColor]] forState: UIControlStateNormal];
    [self.seriesNonMetalButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory nonMetalColor]] forState: UIControlStateNormal];
    [self.seriesTransactinidesButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory transactinideColor]] forState: UIControlStateNormal];
    [self.seriesTransitionMetalButton setTitleColor: [UIColor reverseColorOf: [XTRColorFactory transitionMetalColor]] forState: UIControlStateNormal];
}

- (void)populateElementBubbleState: (BOOL) aFlag {
    [XTRPropertiesStore storeElementBubblesState: aFlag];
}

- (void)populateShowTransitionsState: (BOOL) aFlag {
    [XTRPropertiesStore storeShowTranitionsState: aFlag];
}

- (void)populateSplashScreenState: (BOOL) aFlag {
    [XTRPropertiesStore storeSplashScreenState: aFlag];
}

#pragma mark - Action Methods

- (IBAction) setElementBubbleState: (id) sender {
    [self populateElementBubbleState: self.elementBubbleSwitch.on];
}

- (IBAction)setShowTransitionsState: (id)sender {
    [self populateShowTransitionsState: self.showTransitionsBubbleSwitch.on];
}

- (IBAction)setSplashScreenState: (id)sender {
    [self populateSplashScreenState: self.splashScreenSwitch.on];
}

- (IBAction)showColorPicker: (id)sender {
    UIButton *aButton = sender;
    NSString *title = aButton.titleLabel.text;
    XTRColorPickerViewController *colorPicker = [[XTRAppDelegate storyboard] instantiateViewControllerWithIdentifier: NSStringFromClass([XTRColorPickerViewController class])];
    UIColor *aColor = [XTRColorFactory colorForString: title];
    
    colorPicker.popoverPresentationController.permittedArrowDirections = UIPopoverArrowDirectionDown;
    colorPicker.modalPresentationStyle = UIModalPresentationPopover;
    colorPicker.popoverPresentationController.sourceView = sender;
    colorPicker.popoverPresentationController.sourceRect = CGRectMake(100, 17, 5, 5);
    colorPicker.preferredContentSize = CGSizeMake(270, 175);

    [self presentViewController:colorPicker animated:YES completion:nil];
    
    colorPicker.colorTitle.text = title;
    colorPicker.previewView.backgroundColor = aColor;
    
    [colorPicker presetSlidersWithColor: aColor];
}

- (IBAction)resetPreferences: (id)sender {
    [XTRPropertiesStore resetPreferences];
    
    [self populateSeriesColors];
    [self populateElementBubbleState: YES];
    [self populateShowTransitionsState: YES];
    [self populateSplashScreenState: YES];
    
    [[NSNotificationCenter defaultCenter] postNotificationName: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
    self.elementBubbleSwitch.on = YES;
    self.showTransitionsBubbleSwitch.on = YES;
    self.splashScreenSwitch.on = YES;
}

#pragma mark - View Management Methods

- (BOOL)shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation)interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(colorSelected:) name: NOTIFICATION_COLOR_SELECTED object: nil];
    
    [self loadDocument: @"Credits.rtf" inView:self.webView];
    [self loadUserDefaults];
    [self populateSeriesColors];
    
    self.appNameLabel.text = XTRVersionChecker.appNameString;
    self.versionLabel.text = XTRVersionChecker.appVersionString;
    self.cpyRightLabel.text = XTRVersionChecker.copywriteString;
}

#pragma mark - Memory Management Methods

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
    self.appNameLabel = nil;
    self.cpyRightLabel = nil;
    self.elementBubbleSwitch = nil;
    self.seriesActinideButton = nil;
    self.seriesAlkaliEarthMetalButton = nil;
    self.seriesAlkaliMetalButton = nil;
    self.seriesHalogenButton = nil;
    self.seriesLanthanideButton = nil;
    self.seriesMetalButton = nil;
    self.seriesNobleGasButton = nil;
    self.seriesNonMetalButton = nil;
    self.seriesTransactinidesButton = nil;
    self.seriesTransitionMetalButton = nil;
    self.showTransitionsBubbleSwitch = nil;
    self.splashScreenSwitch = nil;
    self.versionLabel = nil;
    self.webView.delegate = nil;
    self.webView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRPreferencesViewController.swift
//
//  XTRPreferencesViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/22/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

import WebKit

class XTRPreferencesViewController : UIViewController {
    
    @IBOutlet var appNameLabel : UILabel!
    @IBOutlet var cpyRightLabel : UILabel!
    
    @IBOutlet var seriesActinideButton : UIButton!
    @IBOutlet var seriesAlkaliEarthMetalButton : UIButton!
    @IBOutlet var seriesAlkaliMetalButton : UIButton!
    @IBOutlet var seriesHalogenButton : UIButton!
    @IBOutlet var seriesLanthanideButton : UIButton!
    @IBOutlet var seriesMetalButton : UIButton!
    @IBOutlet var seriesNobleGasButton : UIButton!
    @IBOutlet var seriesNonMetalButton : UIButton!
    @IBOutlet var seriesTransactinidesButton : UIButton!
    @IBOutlet var seriesTransitionMetalButton : UIButton!
    
    @IBOutlet var versionLabel : UILabel!
    @IBOutlet var elementBubbleSwitch : UISwitch!
    @IBOutlet var showTransitionsBubbleSwitch : UISwitch!
    @IBOutlet var splashScreenSwitch : UISwitch!
    @IBOutlet var webView : WKWebView!
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    @objc func colorSelected(notification: Notification) {
        guard let anObject : [String : AnyObject] = notification.object as! [String : AnyObject]? else { return }
        
        let redComponent = anObject[ColorComponent.red] as! NSNumber
        let greenComponent = anObject[ColorComponent.green] as! NSNumber
        let blueComponent = anObject[ColorComponent.blue] as! NSNumber
        let alphaComponent = anObject[ColorComponent.alpha] as! NSNumber
        
        let seriesColorKey = anObject[SERIES_COLOR_KEY] as! String
        let aColor = UIColor(red: CGFloat(redComponent.floatValue), green: CGFloat(greenComponent.floatValue), blue: CGFloat(blueComponent.floatValue), alpha: CGFloat(alphaComponent.floatValue))
        let colorData = NSKeyedArchiver.archivedData(withRootObject: aColor)
        
        XTRPropertiesStore.setColorData(colorData, key: seriesColorKey)
        
        if seriesColorKey == ElementSeries.actinide {
            updateSeriesButtonProperties(seriesActinideButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.alkaliEarthMetal {
            updateSeriesButtonProperties(seriesAlkaliEarthMetalButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.alkaliMetal {
            updateSeriesButtonProperties(seriesAlkaliMetalButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.halogen {
            updateSeriesButtonProperties(seriesHalogenButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.lanthanide {
            updateSeriesButtonProperties(seriesLanthanideButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.metal {
            updateSeriesButtonProperties(seriesMetalButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.nobleGas {
            updateSeriesButtonProperties(seriesNobleGasButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.nonMetal {
            updateSeriesButtonProperties(seriesNonMetalButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.transactinides {
            updateSeriesButtonProperties(seriesTransactinidesButton, aColor: aColor)
        } else if seriesColorKey == ElementSeries.transitionMetal {
            updateSeriesButtonProperties(seriesTransitionMetalButton, aColor: aColor)
        }
        
        NotificationCenter.default.post(name: .seriesColorChangedNotification, object: nil)
        
        dismiss(animated: true, completion: nil)
    }
    
    func loadDocument(_ documentName: String, inView: WKWebView) {
        guard let aPath = Bundle(for: XTRElement.classForCoder()).path(forResource: documentName, ofType: nil) else { return }
        
        webView.load(URLRequest(url: URL(fileURLWithPath: aPath)))
    }
    
    func loadUserDefaults() {
        elementBubbleSwitch.isOn = XTRPropertiesStore.elementBubblesState
        showTransitionsBubbleSwitch.isOn = XTRPropertiesStore.showTransitionsState
        splashScreenSwitch.isOn = XTRPropertiesStore.splashScreenState
    }
    
    func updateSeriesButtonProperties(_ aButton: UIButton, aColor: UIColor) {
        aButton.backgroundColor = aColor
        aButton.setTitleColor(aColor.inverseColor(), for: UIControlState())
    }
    
    func populateSeriesColors() {
        updateSeriesButtonProperties(seriesActinideButton, aColor: XTRColorFactory.actinideColor)
        updateSeriesButtonProperties(seriesAlkaliEarthMetalButton, aColor: XTRColorFactory.alkaliEarthMetalColor)
        updateSeriesButtonProperties(seriesAlkaliMetalButton, aColor: XTRColorFactory.alkaliMetalColor)
        updateSeriesButtonProperties(seriesHalogenButton, aColor: XTRColorFactory.halogenColor)
        updateSeriesButtonProperties(seriesLanthanideButton, aColor: XTRColorFactory.lanthanideColor)
        updateSeriesButtonProperties(seriesMetalButton, aColor: XTRColorFactory.metalColor)
        updateSeriesButtonProperties(seriesNobleGasButton, aColor: XTRColorFactory.nobleGasColor)
        updateSeriesButtonProperties(seriesNonMetalButton, aColor: XTRColorFactory.nonMetalColor)
        updateSeriesButtonProperties(seriesTransactinidesButton, aColor: XTRColorFactory.transactinideColor)
        updateSeriesButtonProperties(seriesTransitionMetalButton, aColor: XTRColorFactory.transitionMetalColor)
    }
    
    func populateElementBubbleState(_ aFlag: Bool)  {
        XTRPropertiesStore.elementBubblesState = aFlag
    }
    
    func populateShowTransitionsState(_ aFlag: Bool) {
        XTRPropertiesStore.showTransitionsState = aFlag
    }
    
    func populateSplashScreenState(_ aFlag: Bool) {
        XTRPropertiesStore.splashScreenState = aFlag
    }
    
    // MARK: - Action Methods
    
    @IBAction func setElementBubbleState(_ sender: UISwitch) {
        populateElementBubbleState(elementBubbleSwitch.isOn)
    }
    
    @IBAction func setShowTransitionsState(_ sender: UISwitch) {
        populateShowTransitionsState(showTransitionsBubbleSwitch.isOn)
    }
    
    @IBAction func setSplashScreenState(_ sender: UISwitch) {
        populateSplashScreenState(splashScreenSwitch.isOn)
    }
    
    @IBAction func resetPreferences(_ sender: UIButton) {
        XTRPropertiesStore.resetPreferences()
        
        populateSeriesColors()
        populateElementBubbleState(true)
        populateShowTransitionsState(true)
        populateSplashScreenState(true)
        
        elementBubbleSwitch.isOn = true
        showTransitionsBubbleSwitch.isOn = true
        splashScreenSwitch.isOn = true
        
        NotificationCenter.default.post(name: .seriesColorChangedNotification, object: nil)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let destController = segue.destination as! XTRColorPickerViewController
        let seriesName = (sender as! UIButton).titleLabel!.text!
        
        destController.seriesName = seriesName
        destController.preferredContentSize = CGSize(width: 270, height: 270)
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(self, selector: #selector(XTRPreferencesViewController.colorSelected(notification:)), name: .colorSelectedNotification, object: nil)
        
        loadDocument("Credits.rtf", inView: webView)
        loadUserDefaults()
        populateSeriesColors()
        
        appNameLabel.text = Bundle.main.appNameString
        versionLabel.text = Bundle.main.appVersionString
        cpyRightLabel.text = Bundle.main.copywriteString
        navigationController?.navigationBar.prefersLargeTitles = true
    }
    
    override func viewWillAppear(_ animated: Bool)  {
        super.viewWillAppear(animated)
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .colorSelectedNotification, object: nil)
        
        appNameLabel = nil
        cpyRightLabel = nil
        elementBubbleSwitch = nil
        seriesActinideButton = nil
        seriesAlkaliEarthMetalButton = nil
        seriesAlkaliMetalButton = nil
        seriesHalogenButton = nil
        seriesLanthanideButton = nil
        seriesMetalButton = nil
        seriesNobleGasButton = nil
        seriesNonMetalButton = nil
        seriesTransactinidesButton = nil
        seriesTransitionMetalButton = nil
        showTransitionsBubbleSwitch = nil
        splashScreenSwitch = nil
        versionLabel = nil
        webView = nil
    }
    
}
